### ISSUE TITLE
Cannot connect to server X

### ISSUE THREAD

**Issue by sanchobouillant** [KnpLabs/snappy#89]
Hi,

When i use wkhtmltojpg from the command line it work but when i use it from my application it doesn't. That why i think it come from KnpSnappy.

The error:

```
The exit status code '1' says something went wrong:
stderr: "wkhtmltopdf: cannot connect to X server 
"
stdout: ""
command: /bin/wkhtmltopdf --lowquality --encoding 'utf-8' '/tmp/knp_snappy535a8764551075.01876888.html' '/tmp/knp_snappy535a87645513f8.46100863.pdf'.
```
```
wkhtmltopdf --version
Name:
  wkhtmltopdf 0.12.1-dev
```

Any ideas ?

**Comment by vurcease** [comment]
The virtual x-server seems to be the popular answer, but it's an unsavory option.  I'm using Vagrant to power a Fedora development environment, and I've scripted the yum package manager to install wkhtmltopdf.  I'm running into the x-server requirement as well.

I've added my voice to this bug (https://bugzilla.redhat.com/show_bug.cgi?id=1140279) in hopes the version of wkhtmltopdf that does NOT require x-server can be added to the package.

**Comment by akovalyov** [comment]
Accroding to https://github.com/wkhtmltopdf/wkhtmltopdf/blob/7ffec76429e5db82f24293ac441efe2a9f2bb1d4/src/pdf/pdfarguments.cc#L216, it works without xvfb by default.

**Comment by spackmat** [comment]
In my case the problem was caused by the binary that comes from the Ubuntu repository (for now this is 0.12.4 on my 18.04 dev box). This binary misses the patched qt. So I uninstalled the Distro package and installed the current official package from the Website of wkhtmltopdf and installed that:

1. Do a `wkhtmltopdf -V` and observe the version number without any note on a patched qt. In my case this was "wkhtmltopdf 0.12.4".
2. Uninstall the distribution package of wkhtmltopdf, on my Ubuntu box like this: `sudo apt purge wkhtmltopdf`
3. Go to https://wkhtmltopdf.org/downloads.html and copy the link to the right package for you.
4. Download that package: `wget https://downloads.wkhtmltopdf.org/0.12/0.12.5/wkhtmltox_0.12.5-1.bionic_amd64.deb`
5. Install that package, on Ubuntu like this: `sudo dpkg -i wkhtmltox_0.12.5-1.bionic_amd64.deb`
6. In my case, the package "xfonts-75dpi" was missing. So do a ` sudo apt install -f` to install the missing packages.
7. In my case, the binary is now in `/usr/local/bin/wkhtmltopdf` instead of `/usr/bin/wkhtmltopdf`, so reflect that in your Snappy config.
8. Do a `wkhtmltopdf -V` again and observe the version string with the note on the patched qt, in my case this was "wkhtmltopdf 0.12.5 (with patched qt)".

And with that (official) binary with patched qt, the error has gone away and Snappy renders its PDF files nicely. Should be documented at a prominent position that some distribution binarys of wkhtmltopdf have that problem and how to solve that.

By the way did the not patched wkhtmltopdf work like a charm for me, when I called it directly via CLI, but had that nasty error message when called via Snappy. I don't know why. On my (shared) hosting (running CentOS 7), the installed wkhtmltopdf also shows this error when called via CLI, which makes more sense to me.
